### PR TITLE
.Net: Support DI for Text Search Services

### LINKED ISSUE #6724: .Net: Support DI for Text Search Services
- [x] Initial check-in #9026 
- [x] Comment to be addressed https://github.com/microsoft/semantic-kernel/pull/9026#discussion_r1781330298

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Initial commit for #6724 

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="ITextSearch"/> on.</param>
+    /// <param name="textSearch">Instance of <see cref="ITextSearch"/> to register.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    public static IServiceCollection AddTextSearch(
+        this IServiceCollection services,
+        ITextSearch textSearch,
+        string? serviceId = default)
+    {
+        ITextSearch Factory(IServiceProvider serviceProvider, object? _) => textSearch;
+        services.AddKeyedSingleton<ITextSearch>(serviceId, (Func<IServiceProvider, object?, ITextSearch>)Factory);
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:27]: Consider this overload, which allows you to just pass the required instance without needing a factory when registering a singleton:
https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.servicecollectionserviceextensions.addkeyedsingleton?view=net-8.0#microsoft-extensions-dependencyinjection-servicecollectionserviceextensions-addkeyedsingleton-1(microsoft-extensions-dependencyinjection-iservicecollection-system-object-0)

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        // cannot make assumptions about how dependent services are being managed.
+        services.AddKeyedTransient<VectorStoreTextSearch<TRecord>>(
+            serviceId,
+            (sp, obj) =>
+            {
+                var stringMapper = sp.GetRequiredService<ITextSearchStringMapper>();
+                var resultMapper = sp.GetRequiredService<ITextSearchResultMapper>();
+                var options = sp.GetService<VectorStoreTextSearchOptions>();
+
+                var vectorizableTextSearch = sp.GetService<IVectorizableTextSearch<TRecord>>();
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:68]: Should you be able to say which instance of IVectorizableTextSearch you want, if more than one is registered, e.g. via service id?

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        services.AddKeyedSingleton<VectorStoreTextSearch<TRecord>>(serviceId, (Func<IServiceProvider, object?, VectorStoreTextSearch<TRecord>>)Factory);
+        return services;
+    }
+
+    /// <summary>
+    /// Register a <see cref="VectorStoreTextSearch{TRecord}"/> instance with the specified service ID.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="ITextSearch"/> on.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    public static IServiceCollection AddVectorStoreTextSearch<TRecord>(
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:53]: Consider an overload where we take the mappers and options as input, but resolve the `IVectorizableTextSearch` from DI.  I would expect that to be a common scenario, e.g. the user is using the existing helpers for VectorStore to register an instance of IVectorizableTextSearch, and now they want to consume that and provide some specific options for how to consume it.

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+/// </summary>
+public static class TextSearchServiceCollectionExtensions
+{
+    /// <summary>
+    /// Register an <see cref="ITextSearch"/> instance with the specified service ID.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="ITextSearch"/> on.</param>
+    /// <param name="textSearch">Instance of <see cref="ITextSearch"/> to register.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    public static IServiceCollection AddTextSearch(
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:21]: I'm not super convinced about the value of the first two overloads here.  I think what's more useful would be something like AddBingSearch that just takes the options for that.  After all someone can just call AddKeyedSingleton themselves here.  We are not really doing anything extra.

**Review by westey-m — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs (hunk 4 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:21]: Yeah, I agree. I'll move these to the Web plugin

File: dotnet/src/Plugins/Plugins.UnitTests/Web/Bing/BingTextSearchTests.cs [hunk 5]
```diff
@@ -25,6 +27,20 @@ public BingTextSearchTests()
...
+    {
+        // Arrange
+        var builder = Kernel.CreateBuilder();
+
+        // Act
+        builder.AddBingTextSearch(apiKey: "ApiKey");
+        var kernel = builder.Build();
+
+        // Assert
+        Assert.NotNull(kernel.Services.GetRequiredService<ITextSearch>());
```
Comment (SergeyMenshykh) [inline dotnet/src/Plugins/Plugins.UnitTests/Web/Bing/BingTextSearchTests.cs:41]: nit: The `GetRequiredService` method will never return null. As I understand it, it either returns the requested service or throws an exception. Perhaps you can check if the service returned by the method is of type `BingTextSearch`.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs (hunk 2 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Core/Data/TextSearch/TextSearchServiceCollectionExtensions.cs:41]: I'm adding this comment to the task as something I need to address. I want to play around with having multiple vector stores registered to see what makes sense to do.

**Pull request merged**